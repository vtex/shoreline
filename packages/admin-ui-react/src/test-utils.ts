import type { ReactElement } from 'react'
import type { RenderOptions } from '@testing-library/react'
import { render as baseRender } from '@testing-library/react'
import { createSystem } from './system'

const [ThemeProvider] = createSystem()

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

export { render }
