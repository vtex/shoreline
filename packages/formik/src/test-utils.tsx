import '@testing-library/jest-dom/extend-expect'
import React, { useRef } from 'react'
import { axe, toHaveNoViolations } from 'jest-axe'
import type { RenderOptions } from '@testing-library/react'
import {
  render as baseRender,
  waitFor,
  screen,
  fireEvent,
} from '@testing-library/react'
import { createSystem } from '@vtex/admin-ui-react'
import type { ReactElement } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'

expect.extend(toHaveNoViolations)

const [ThemeProvider] = createSystem()

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

/**
 * HOC to wrap a component with its state hook
 * @example
 * // Cmp has the useCmp state hook
 * const StatefulCpm = withState(Cmp, () => useCmpState({ isSelected: false }))
 */
function withState(Component: any, useStateHook: () => any) {
  function StatefulComponent(props: any) {
    // Avoids an invalid hook call
    const ref = useRef(useStateHook())

    return <Component state={ref.current} {...props} />
  }

  StatefulComponent.displayName = Component.displayName

  return StatefulComponent
}

function jestMatchMedia() {
  /**
   * ðŸš§ Workaround for window.match media
   * @see https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
   */
  return Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
}

export {
  render,
  withState,
  axe,
  jestMatchMedia,
  renderHook,
  act,
  waitFor,
  screen,
  fireEvent,
}
