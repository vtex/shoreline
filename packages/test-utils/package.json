{
  "name": "@vtex/shoreline-test-utils",
  "version": "1.0.32",
  "description": "Shoreline test utilities",
  "main": "./dist/index.mjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.mts",
  "repository": "vtex/shoreline",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "types": "./dist/index.d.mts"
    },
    "./setup": {
      "import": "./dist/setup.mjs",
      "types": "./dist/setup.d.mts"
    }
  },
  "engines": {
    "node": ">=20"
  },
  "scripts": {
    "prebuild": "rm -rf dist",
    "dev": "tsup --watch",
    "build": "npm run prebuild && tsup"
  },
  "peerDependencies": {
    "vitest": "0.x"
  },
  "devDependencies": {
    "vitest": "0.34.6"
  },
  "dependencies": {
    "@testing-library/dom": "9.3.4",
    "@testing-library/jest-dom": "6.4.6",
    "@testing-library/react": "14.3.1",
    "@testing-library/user-event": "14.5.2"
  }
}
