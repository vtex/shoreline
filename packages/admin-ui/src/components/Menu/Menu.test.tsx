import React from 'react'
import { render, axe, withState } from '../../test-utils'
import 'mutationobserver-shim'

import {
  MenuSeparator,
  Menu,
  MenuButton,
  MenuList,
  MenuItem,
  useMenuState,
} from './index'

global.MutationObserver = window.MutationObserver

const StatefulMenu = withState(Menu, () =>
  useMenuState({ visible: true, baseId: 'id' })
)

describe('Menu', () => {
  it('should not have a11y violations', async () => {
    const { container } = render(
      <StatefulMenu>
        <MenuButton>Open</MenuButton>
        <MenuList
          data-testid="menu"
          aria-label="menu label"
          csx={{
            bg: 'coral',
          }}
        >
          <MenuItem>Download</MenuItem>
          <MenuItem>Link to</MenuItem>
          <MenuItem>Favorite</MenuItem>
          <MenuSeparator />
          <MenuItem>Delete</MenuItem>
        </MenuList>
      </StatefulMenu>
    )

    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
