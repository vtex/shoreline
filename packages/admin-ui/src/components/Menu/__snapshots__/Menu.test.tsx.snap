// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should match snapshot hidden 1`] = `
<DocumentFragment>
  

.emotion-3 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  height: 40px;
  width: auto;
  padding-left: 0.75rem;
  padding-right: 0.5rem;
  color: #FFFFFF;
  background: hsla(222, 63%, 43%, 1);
}

.emotion-3:disabled {
  background: hsla(0, 0%, 90%, 1);
  color: hsla(0, 0%, 45%, 1);
}

.emotion-3:hover {
  color: #FFFFFF;
  background: hsla(222, 63%, 34%, 1);
}

.emotion-3:active {
  color: #FFFFFF;
  background: hsla(222, 63%, 25%, 1);
}

.emotion-3:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.125rem hsla(222, 64%, 84%, 1);
}

.emotion-3 svg {
  margin: 0.25rem;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  width: 100%;
  margin: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
}

.emotion-1 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  visibility: visible;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.emotion-0 {
  width: 1.25rem;
  height: 1.25rem;
  min-height: 1.25rem;
  min-width: 1.25rem;
}

.emotion-13 {
  outline: none;
  z-index: 999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: coral;
  padding: 0.75rem;
  min-width: 18px;
  border: 1px solid hsla(0, 0%, 87%, 1);
  border-radius: 3px;
  box-shadow: 0rem 0rem 1rem 0rem rgba(0,0,0,0.10);
}

.emotion-5 {
  margin-top: 0.125rem;
  margin-bottom: 0.125rem;
  padding-left: 0.25rem;
  padding-right: 0.25rem;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 475;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0071em;
  border: none;
  text-transform: initial;
  width: 100%;
  height: 32px;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  background: transparent;
  color: hsla(222, 63%, 43%, 1);
}

.emotion-5 div {
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
}

.emotion-5 svg {
  margin: 0.25rem;
  width: 20px;
  height: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-left: 0rem;
  margin-right: 0.5rem;
}

.emotion-5:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-5:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.125rem hsla(222, 64%, 84%, 1);
}

.emotion-5:hover {
  color: hsla(222, 63%, 43%, 1);
  background: hsla(224, 65%, 97%, 1);
}

.emotion-5:active {
  color: hsla(222, 63%, 43%, 1);
  background: hsla(222, 62%, 92%, 1);
}

.emotion-5:disabled {
  color: hsla(0, 0%, 45%, 1);
  background: hsla(0, 0%, 90%, 1);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-10 {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  border-bottom: 1px solid hsla(0, 0%, 87%, 1);
  width: calc(100% -0.75rem);
  margin-left: -0.75rem;
  margin-right: -0.75rem;
  outline: none;
}

<div
    class="emotion-14"
  >
    <button
      aria-controls="id"
      aria-expanded="false"
      aria-haspopup="menu"
      class="emotion-3"
      type="button"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          <svg
            class="emotion-0"
            fill="currentColor"
            height="1.25rem"
            viewBox="0 0 256 256"
            width="1.25rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="16rem"
              width="16rem"
            />
            <polyline
              fill="none"
              points="208 96 128 176 48 96"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="18"
            />
          </svg>
           Open
        </div>
      </div>
    </button>
    <div
      aria-label="menu label"
      aria-orientation="vertical"
      class="emotion-13"
      data-dialog="true"
      data-testid="menu"
      hidden=""
      id="id"
      role="menu"
      style="display: none; position: fixed; left: 100%; top: 100%;"
      tabindex="-1"
    >
      <button
        class="emotion-5"
        id="id-1"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Download
        </div>
      </button>
      <button
        class="emotion-5"
        id="id-2"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Link to
        </div>
      </button>
      <button
        class="emotion-5"
        id="id-3"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Favorite
        </div>
      </button>
      <hr
        aria-orientation="horizontal"
        class="emotion-10"
        role="separator"
      />
      <button
        class="emotion-5"
        id="id-4"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Delete
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Menu should match snapshot visible 1`] = `
<DocumentFragment>
  

.emotion-3 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  height: 40px;
  width: auto;
  padding-left: 0.75rem;
  padding-right: 0.5rem;
  color: #FFFFFF;
  background: hsla(222, 63%, 43%, 1);
}

.emotion-3:disabled {
  background: hsla(0, 0%, 90%, 1);
  color: hsla(0, 0%, 45%, 1);
}

.emotion-3:hover {
  color: #FFFFFF;
  background: hsla(222, 63%, 34%, 1);
}

.emotion-3:active {
  color: #FFFFFF;
  background: hsla(222, 63%, 25%, 1);
}

.emotion-3:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.125rem hsla(222, 64%, 84%, 1);
}

.emotion-3 svg {
  margin: 0.25rem;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  width: 100%;
  margin: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
}

.emotion-1 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 600;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0125em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  visibility: visible;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.emotion-0 {
  width: 1.25rem;
  height: 1.25rem;
  min-height: 1.25rem;
  min-width: 1.25rem;
}

.emotion-5 {
  margin-top: 0.125rem;
  margin-bottom: 0.125rem;
  padding-left: 0.25rem;
  padding-right: 0.25rem;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 475;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.0071em;
  border: none;
  text-transform: initial;
  width: 100%;
  height: 32px;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  background: transparent;
  color: hsla(222, 63%, 43%, 1);
}

.emotion-5 div {
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
}

.emotion-5 svg {
  margin: 0.25rem;
  width: 20px;
  height: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-left: 0rem;
  margin-right: 0.5rem;
}

.emotion-5:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-5:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.125rem hsla(222, 64%, 84%, 1);
}

.emotion-5:hover {
  color: hsla(222, 63%, 43%, 1);
  background: hsla(224, 65%, 97%, 1);
}

.emotion-5:active {
  color: hsla(222, 63%, 43%, 1);
  background: hsla(222, 62%, 92%, 1);
}

.emotion-5:disabled {
  color: hsla(0, 0%, 45%, 1);
  background: hsla(0, 0%, 90%, 1);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-10 {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  border-bottom: 1px solid hsla(0, 0%, 87%, 1);
  width: calc(100% -0.75rem);
  margin-left: -0.75rem;
  margin-right: -0.75rem;
  outline: none;
}

.emotion-13 {
  outline: none;
  z-index: 999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #FFFFFF;
  padding: 0.75rem;
  min-width: 18px;
  border: 1px solid hsla(0, 0%, 87%, 1);
  border-radius: 3px;
  box-shadow: 0rem 0rem 1rem 0rem rgba(0,0,0,0.10);
}

<div
    class="emotion-14"
  >
    <button
      aria-controls="id"
      aria-expanded="true"
      aria-haspopup="menu"
      class="emotion-3"
      type="button"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          <svg
            class="emotion-0"
            fill="currentColor"
            height="1.25rem"
            viewBox="0 0 256 256"
            width="1.25rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="16rem"
              width="16rem"
            />
            <polyline
              fill="none"
              points="208 96 128 176 48 96"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="18"
            />
          </svg>
           Open
        </div>
      </div>
    </button>
    <div
      aria-label="menu label"
      aria-orientation="vertical"
      class="emotion-13"
      data-dialog="true"
      data-testid="menu"
      id="id"
      role="menu"
      style="position: fixed; left: 100%; top: 100%;"
      tabindex="-1"
    >
      <button
        class="emotion-5"
        id="id-1"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Download
        </div>
      </button>
      <button
        class="emotion-5"
        id="id-2"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Link to
        </div>
      </button>
      <button
        class="emotion-5"
        id="id-3"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Favorite
        </div>
      </button>
      <hr
        aria-orientation="horizontal"
        class="emotion-10"
        role="separator"
      />
      <button
        class="emotion-5"
        id="id-4"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-4"
        >
           Delete
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
